File: sdk/node/src/middleware.ts
import type { Context } from "./runtime";

/**
 * Middleware intercepts handler execution. Call `next()` to invoke the
 * next middleware or the terminal handler. Middleware is applied in FIFO
 * order: the first registered middleware is the outermost.
 */
export type Middleware = (ctx: Context, next: () => Promise<any>) => Promise<any>;

/**
 * Returns a middleware that logs job ID, topic, and duration for every request.
 *
 * @param logger - Logger to use. Defaults to `console`.
 */
export function loggingMiddleware(logger: { info: (...args: any[]) => void } = console): Middleware {
  return async (ctx: Context, next: () => Promise<any>): Promise<any> => {
    const start = Date.now();
    try {
      const result = await next();
      const elapsed = Date.now() - start;
      logger.info(`middleware: job=${ctx.jobId} topic=${ctx.job?.topic ?? "?"} duration=${elapsed}ms ok`);
      return result;
    } catch (err) {
      const elapsed = Date.now() - start;
      logger.info(`middleware: job=${ctx.jobId} topic=${ctx.job?.topic ?? "?"} duration=${elapsed}ms error=${err}`);
      throw err;
    }
  };
}

File: sdk/cpp/src/middleware.cpp
#include "cap/middleware.h"

#include <chrono>
#include <iostream>

namespace cap {

Handler ApplyMiddleware(Handler base, const std::vector<Middleware>& middlewares) {
  Handler handler = std::move(base);
  for (auto it = middlewares.rbegin(); it != middlewares.rend(); ++it) {
    handler = (*it)(std::move(handler));
  }
  return handler;
}

Middleware LoggingMiddleware() {
  return [](Handler next) -> Handler {
    return [next = std::move(next)](const Context& ctx,
                                    const cordum::agent::v1::JobRequest& req)
               -> std::unique_ptr<cordum::agent::v1::JobResult> {
      auto start = std::chrono::steady_clock::now();
      auto result = next(ctx, req);
      auto end = std::chrono::steady_clock::now();
      auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
      std::cout << "[cap] job=" << req.job_id() << " duration=" << ms << "ms" << std::endl;
      return result;
    };
  };
}

}  // namespace cap
